### ISSUE TITLE
Fix `Todo` type inferred when subscripting an intersection

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ty#2072]
```py
from typing import Sequence

class C: ...

def f(x: Sequence[int]):
    if isinstance(x, C):
        reveal_type(x[0])  # revealed: @Todo
```

We should infer `int` there (and, same as our current behaviour, we should not emit a diagnostic).

This was previously attempted in https://github.com/astral-sh/ruff/pull/18846 but then reverted in https://github.com/astral-sh/ruff/pull/18920

**Comment by AlexWaygood** [comment]
(@charliermarsh I think I previously asked you to hold off on this one because it would conflict with https://github.com/astral-sh/ruff/pull/22035 -- but that one's now merged, so feel free to have at it!)
